packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol

```diff
@@ -358,7 +361,9 @@ abstract contract CrossDomainMessenger is
         );
 
         xDomainMsgSender = _sender;
-        bool success = SafeCall.callWithMinGas(_target, _minGasLimit, _value, _message);
+        // Add an extra 30_000 reserved gas to retain after the safe call to ensure that there
+        // is enough gas to finish the execution within this context afterwards.
+        bool success = SafeCall.callWithMinGas(_target, _minGasLimit, _value, 30_000, _message);
```

Comment:
I propose that we: 

1. Revert #5108. 
2. Rebase on #5444 once it's in.

Combine the old gas buffer check with the new reentrancy protection check to be:

```diff
    if (xDomainMsgSender != Constants.DEFAULT_L2_SENDER 
+	|| gasleft() >= _minGasLimit + RELAY_GAS_REQUIRED
    ) {
            failedMessages[versionedHash] = true;
            emit FailedRelayedMessage(versionedHash);
            return;
        }
```

Reply:
I do like this direction but am unsure of exactly how it would help to unstick a message as we do not allow for passing in a different `minGasLimit`. The old contracts do allow for this, there is a different function `replayMessage`. This at least would give us better monitoring into the issue, we would be able to observe the set of failed messages

Reply:
by `RELAY_GAS_REQUIRED` do you mean `MIN_GAS_CONSTANT_OVERHEAD`?

Reply:
> ... we do not allow for passing in a different `minGasLimit`

We don't allow modifying `minGasLimit`, but `SafeCall.call()` can accept `gasleft()` as its `_gas` [argument](https://github.com/ethereum-optimism/optimism/blob/d460e7670923a115fe2c3c3af14914df8394b849/packages/contracts-bedrock/contracts/libraries/SafeCall.sol#L17-L26) which it will forward.

`RELAY_GAS_REQUIRED` was the constant value in a previous version of the contract which was "the amount of gas held in reserve to guarantee that relay execution completes".

https://github.com/ethereum-optimism/optimism/blob/608ebc9a8a597d57acc71d4740fe9b1980357380/packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol#L370-L377

There were other issues with the implementation (not accounting for the SSTORE to set `xDomainMessageSender`), but we can address that.

Reply:
I think the pattern we need to avoid is any logic that triggers a revert based on the `gasleft()`.

Reply:
I agree that not reverting here is the safer route. I've rebased the PR on top of the other XDM changes and changed it up.

Reply:
I am starting to feel reluctant about how large of a diff this is

Reply:
Think we're in a good place to resolve this convo

Reply:
> Think we're in a good place to resolve this convo

Oh, didn't realize this was my comment to begin with. Resolved.